[cmake/linux] Code coverage

fetzerch opened the pull request:
Now that the tests are working again, I thought it might be a good idea to see what we are testing. 

**Usage:**

```
cmake -DCMAKE_BUILD_TYPE=Coverage <SRC-DIR>/project/cmake`
make coverage
```

Tests are executed as part of the `coverage` or `coverage_xml` target.

**Output:**
![unbenannt](https://cloud.githubusercontent.com/assets/1133994/16707850/1099b134-45de-11e6-928f-f166329fdea2.PNG)

I've triggered a build: http://jenkins.kodi.tv/view/Linux/job/LINUX-64/8340 to see if we have the necessary tools installed on Jenkins. ([Cobertura Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Cobertura+Plugin) on the master and `lcov` and `gcovr` on the Linux slave(s)).
Maybe @Memphiz or @wsnipex can help here.

**Known Issue:**
There's one known issue: I had to disable library wrapping in the `Coverage` build type, because there are errors such as:

```
cc: error: __gcov0.-Wl,-wrap,calloc: No such file or directory
```

While this doesn't affect the reports much, it might lead to surprising effects when you're running Kodi built with the `Coverage` type. Just something to be aware of for the moment.
Maybe @notspiff can help to setup `wrapper.def` for this.

**Other platforms:**
This is Linux only for the moment, if anyone is interested to get coverage to work on Windows or OSX, I can assist.

fetzerch committed:
[cmake/linux] Code Coverage

Adds a FindCoverage.cmake script that generates code coverages reports
using LCOV. As coverage requires to compile with different options,
a 'Coverage' build type has been introduced.

Usage:
    cmake -DCMAKE_BUILD_TYPE=Coverage <KODI_SRC>/project/cmake \
          && make -j$(nproc) coverage

Known issues: Doesn't work with wrapped libraries (i.g. libcpluff).
Therefore these are compiled unwrapped and the output behaves
differently compared to Debug or Release build!

fetzerch committed:
[cddb] Drop non UTF-8 character

fetzerch commented on xbmc/network/cddb.cpp:
```diff
@@ -219,8 +219,8 @@ bool Xcddb::queryCDinfo(CCdInfo* pInfo, int inexact_list_select)
 
   //##########################################################
   // Read the data from cddb
-  Recv(false); //erstmal den M�ll abholen
```
:laughing: unfortunately gcovr doesn't like non UTF-8 chars.

Memphiz replied:
mhh dropping the comment without replacing it with something english? By intention?

fritsch commented on xbmc/network/cddb.cpp:
```diff
@@ -219,8 +219,8 @@ bool Xcddb::queryCDinfo(CCdInfo* pInfo, int inexact_list_select)
   // Read the data from cddb
-  Recv(false); //erstmal den M�ll abholen
-  if ( !Send(read_buffer.c_str()) )
+  Recv(false);
```
`` `
// Clear pending data on our data connection
```

fetzerch commented:
Thx, done.

wsnipex commented:
sadly our jenkins slave on linux is still ubuntu 12.04, which does not provide gcovr...

and I've sent you a small PR that eliminates the need for "make all kodi-test"

wsnipex commented:
installed gcovr and lcov on the jenkins slave

fetzerch committed:
[depends/cmakebuildsys] Pass Configuration to the build

fetzerch committed:
[buildsteps/linux] Run tests with Code Coverage report generation

fetzerch commented:
Thanks @wsnipex. Squashed your changes with a few minor modifications. (COVERAGE_DEPENDS variable, if logic in main CMakeLists.txt)

Next build triggered: http://jenkins.kodi.tv/view/Helpers/job/LINUX-64/8349/

fetzerch committed:
[cmake/linux] Make Code Coverage compatible with lcov < 1.10.

This change can be reverted once there is no need to support
lcov < 1.10 anymore (i.g. on Ubuntu 12.04).

fetzerch commented:
@wsnipex: Seems like there's some incompatibility which causes:

```
geninfo: ERROR: /home/jenkins/workspace/LINUX-64/build/build/api1AddonCallbacks_PVR/CMakeFiles/api1AddonCallbacks_PVR.dir/AddonCallbacksPVR.cpp.gcno: reached unexpected end of file
```
(full log: http://jenkins.kodi.tv/view/Linux/job/LINUX-64/8357/console)

I'm not sure if if can be workarounded without upgrading gcc and lcov. I'll still leave the jenkins bits in place, as they don't hurt. But I fear for now it'll only work locally.

fetzerch commented:
@wsnipex, @hudokkow: Any objections to merging this, even though it doesn't work on jenkins? It might still be useful when working on the tests locally. Currently I don't have time to see if it can be made compatible with 12.04 unfortunately.

hudokkow commented:
Fine by me.
